### ISSUE TITLE
localized keyboard maps with gtkgreet

### ISSUE THREAD

**Issue by mcepl** [openSUSE/openSUSEway#104]
I have this `/etc/greetd/sway-config`:

```ini
# exec "wlgreet --coomand /usr/bin/sway-run.sh ; swaymsg exit"
exec "gtkgreet -l -s /etc/greetd/style.css; swaymsg exit"

bindsym Mod4+shift+e exec swaynag \
	-t warning \
	-m 'What do you want to do?' \
	-b 'Poweroff' 'systemctl poweroff' \
	-b 'Reboot' 'systemctl reboot'

exec /usr/libexec/polkit-gnome-authentication-agent-1

#include /etc/sway/config.d/
```

My problem is that I need my localized keyboard (`cs_CZ.qwerty`) to write correctly password. I have correct keyboard when I start in the text mode, and I can switch to the correct keyboard in the final sway session, but greet login prompt is always in the English keyboard.

My system locale are properly set up:

```sh
stitny:/etc/greetd # localectl status
System Locale: LANG=cs_CZ.UTF-8
    VC Keymap: cz-qwerty
   X11 Layout: cz
    X11 Model: microsoftpro
  X11 Variant: qwerty
  X11 Options: terminate:ctrl_alt_bksp,grp:shift_toggle,ctrl:nocaps,compose:rctrl
stitny:/etc/greetd # 
```

[This message on the upstream list](https://lists.sr.ht/~kennylevinsen/greetd-devel/%3Cfcoed3k3FmTv5o9rtbuc-ZDRYQXLkdxOtM5DXEUrFWiF2J3uCJDv-yGf386hm6ePFLR1Yc5iYrnGOSEzyGAXQZF5ewX6WcRqoOeAZv1Su2o%3D%40pm.me%3E) seems to suggest that the localization is done with configuration file (in our case `/etc/sway/config.d/50-openSUSE.conf`). That configuration unfortunately includes nothing about the localization of the keyboard maps.

**Comment by FilippoBonazziSUSE** [comment]
That's an interesting issue.

### Correct solution
Ideally the greeter should offer a graphical way to switch keyboard maps, since different users of the system may want to use different keymaps to type their usernames and passwords.

### Hack
Since this issue appears before login, it is not correct in principle to fix it by sourcing user configuration. You could of course posit that your system is a single-graphical-user system, and therefore it is acceptable to fix a pre-login graphical environment with a custom user configuration. You could then include `/home/mcepl/.config/sway/config.d/10-input.conf` (or whatever file contains your `xkb_layout` definitions) in `/etc/greetd/sway-config`.

How to do this in a general way for the `openSUSEway` configuration remains an open question...

#### Sad remark
In general, using passwords which rely on a non-US keyboard, in settings where you must type the password by hand on a keyboard you don't control, is a headache.

**Comment by mcepl** [comment]
[sway-config.zip](https://github.com/openSUSE/openSUSEway/files/10954140/sway-config.zip)

Specifically, my problem is that [numbers are on the Czech keyboard in the Shift position](https://en.wikipedia.org/wiki/List_of_QWERTY_keyboard_language_variants#Czech).

**Comment by FilippoBonazziSUSE** [comment]
> Has any progress been made on this bug? 

No, this feature is absent in upstream gtkgreet and I don't believe there is an easy way to work around its absence. I am not sure how you could show different available keymaps, and allow to switch between them, on the gtkgreet login screen without support from the greeter.

HOWEVER, you can of course specify a fixed keyboard layout for the wm you are using to display your gtkgreet. If you know the keyboard layout of your machine, you can set it in that configuration file (`/etc/greetd/sway-config` in openSUSEway).

The [XFCE greetd configuration](https://build.opensuse.org/package/show/openSUSE:Factory/greetd-gtkgreet-xfce-wayland/) (completely unrelated to openSUSEway, btw) uses the `cage` compositor, so you would need to find out how to specify a keyboard layout for it.

**Comment by mcepl** [comment]
On Fri Oct 31, 2025 at 6:15 PM CET, Jean Delvare wrote:
> Well I'm not even asking for a keymap selector. However, the greeter should honor the keyboard mapping setting which was selected at system installation time. The text consoles pick the system default just fine, so I can't imagine it would be difficult for the graphical greeter to do the same.

I think that our only hope is to somehow squeeze into `/etc/greetd/config.toml` some runner for `command` variable, where something would modify `/etc/greetd/sway-config` to use keyboard according to the current locale and then run `sway` with the modified configuration.

**Comment by FilippoBonazziSUSE** [comment]
> Well I'm not even asking for a keymap selector. However, the greeter should honor the keyboard mapping setting which was selected at system installation time. The text consoles pick the system default just fine, so I can't imagine it would be difficult for the graphical greeter to do the same.

I don't like that tone. This is a volunteer hobby project, so your expectation of "I can't imagine it would be difficult" is out of line. What's even funnier is that you are not even using openSUSEway or the openSUSEway greetd configuration, so anything we fix here will not apply to the configuration you are running anyway.

> I think that our only hope is to somehow squeeze into `/etc/greetd/config.toml` some runner for `command` variable, where something would modify `/etc/greetd/sway-config` to use keyboard according to the current locale and then run `sway` with the modified configuration.

After giving the topic some more thought, I came up with something that should work in #219. Can you try out if it works for you? I tested it and it works, but I would like a bigger consensus for it before merging.

**Comment by jdelvare** [comment]
Sorry, I did not mean to sound offensive, I have great respect for hobby open-source projects. I'll delete my messages, which I should not have posted here in the first place anyway, as it turns out my own issue was unrelated to the problem being discussed here.

Hopefully Matej can test your candidate fix.

**Comment by FilippoBonazziSUSE** [comment]
No hard feelings. Thanks for spurring us to revisit an issue that I had forgotten. I'm sure whatever fix we end up going with can then then be taken as inspiration to fix the XFCE greetd configuration (two birds with one stone).

**Comment by bablokb** [comment]
I found this issue because I have the same problem with XFCE. I tried the PR #220 you linked to, but it does not work. The grep-command does what it is expected to do, but the result is a non-functional keyboard. I had to start a rescue system to revert the toml-file again.

**Comment by FilippoBonazziSUSE** [comment]
The [XFCE greetd config](https://build.opensuse.org/package/show/openSUSE:Factory/greetd-gtkgreet-xfce-wayland/) is unrelated to openSUSEway and even runs a different compositor (`cage` instead of `sway`). So any solution we try or reach here can never be directly applicable.

**Comment by bablokb** [comment]
Oh, so I misunderstood your last comment mentioning XFCE. But you are right, you only talked about "inspiration" and not that the solution is similar.
